### PR TITLE
Add 'command' plugin

### PR DESCRIPTION
**Pull request by elcore**
### 1. What does this change do, exactly?

1.  Add the 'command' plugin
2. Register a new Event Hook `CertRenewEvent`
3. Register a new Event Hook `InstanceStartupEvent`

### 2. Please link to the relevant issues.

#1698 

### 3. Which documentation changes (if any) need to be made because of this PR?

None

### 4. Checklist

- [ ] I have written tests and verified that they fail without my change
- [X] I have squashed any insignificant commits
- [X] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I am willing to help maintain this change if there are issues with it later

### CODE REVIEW

**Comment by abiosoft** [comment]
Mind adding a readme to the command repo to indicate how it is meant to be used ?

**Comment by elcore** [comment]
@abiosoft 

> Mind adding a readme to the command repo to indicate how it is meant to be used ?

I am working on it 😄

**Comment by abiosoft** [comment]
I can't help but feel `command` sounds generic. `tlscommand` maybe, considering this is limited to when tls cert has been renewed.

**Comment by elcore** [comment]
> I can't help but feel command sounds generic. tlscommand maybe, considering this is limited to when tls cert has been renewed.

This might change in the near future, it a feature request by Matt H.

**Comment by francislavoie** [comment]
Imo this could (should?) be extended to support multiple different events. I'm thinking more of a `command [event] [cmd] [args]` syntax, where this event specifically would be `certrenew`. Making it more generic would be nice.

Even better, why don't we replace `startup` and `shutdown` with `command startup` and `command shutdown` instead of this being a plugin?

**Comment by elcore** [comment]
>Imo this could (should?) be extended to support multiple different events. I'm thinking more of a command [event] [cmd] [args] syntax, where this event specifically would be certrenew. Making it more generic would be nice.

> Even better, why don't we replace startup and shutdown with command startup and command shutdown instead of this being a plugin?

That is my plan (see above)

**Comment by francislavoie** [comment]
Sorry, maybe I'm blind, but I don't see your plan anywhere?

**Comment by elcore** [comment]
Hello @francislavoie,

> Sorry, maybe I'm blind, but I don't see your plan anywhere?

I have only expressed it vaguely 😄 -- "This might change in the near future, it a feature request by Matt H."

Anyway, I have a "working" local branch

## Syntax

`command [event] [cmd] [args]`

## Example

`command certrenew echo "This is an example"`

File: plugins.go [hunk 1]
```diff
@@ -217,10 +217,12 @@ func RegisterPlugin(name string, plugin Plugin) {
 // EventName represents the name of an event used with event hooks.
 type EventName string
 
-// Define the event names for the startup and shutdown events
+// Define the event names for the startup, shutdown, instancestartup and certrenew events
```
Comment (francislavoie) [inline plugins.go]: Might as well just change this to "names for the various events", it'll get pretty wordy if you keep adding each one to the comment every time you set up a new event 😆

Comment (elcore) [reply plugins.go]: Right 😆

**Comment by elcore** [comment]
I have pushed my local branch 😄

**Comment by francislavoie** [comment]
Awesome @elcore :smiley: 

@mholt IMO this is something that should just be built-into Caddy, I don't think this should only be available via a plugin. I think there is room for plugins to do cool things with events, but just running a command seem like something that should be core. What do you think?

Like I suggested above, I think this should replace the `startup` and `shutdown` directives. I guess for backwards compat they can be kept as aliases to `command startup [args]` and `command shutdown [args]` until later and just marked as deprecated?

**Comment by elcore** [comment]
> IMO this is something that should just be built-into Caddy, I don't think this should only be available via a plugin. [...] What do you think? 

ping @mholt

**Comment by mholt** [comment]
Hi -- sorry, been really busy with other things.

I agree with the others. Let's switch startup and shutdown to use the event hooks, but let's keep those directives as syntactic as sugar. I think we can wrap this all into a standard `run` or `on` directive to run commands on certain events.

Thanks @elcore and others who reviewed!

**Pull request closed without merging**